src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs

```diff
@@ -0,0 +1,173 @@
+                    return MakeCall(node.Syntax, rewrittenReceiver: null, method, arguments, node.Type);
+                }
+                else
+                {
+                    // Case 7:              "nullableLeft..nullableRight"
+                    // translates to:       nullableLeft.HasValue && nullableRight.HasValue ? new Nullable(Range.Create(nullableLeft.GetValueOrDefault(), nullableRight.GetValueOrDefault())) : default
+
+                    // Case 8:              "nullableLeft..right"
+                    // translates to:       nullableLeft.HasValue ? new Nullable(Range.Create(nullableLeft.GetValueOrDefault(), right)) : default
+
+                    // Case 9:              "left..nullableRight"
+                    // translates to:       nullableRight.HasValue ? new Nullable(Range.Create(left, nullableRight.GetValueOrDefault())) : default
+
+                    MethodSymbol method = (MethodSymbol)_compilation.GetWellKnownTypeMember(WellKnownMember.System_Range__Create);
+                    return LiftRangeExpression(node.Syntax, node.Type, method, leftOperand, rightOperand);
+                }
+            }
+        }
+
+        private BoundNode LiftRangeExpression(SyntaxNode syntax, TypeSymbol type, MethodSymbol method, params BoundExpression[] operands)
+        {
+            Debug.Assert(type.IsNullableType());
+            Debug.Assert(operands.Any(operand => operand.Type.IsNullableType()));
+            Debug.Assert(operands.Length == 1 || operands.Length == 2);
+
+            ArrayBuilder<BoundExpression> sideeffects = ArrayBuilder<BoundExpression>.GetInstance();
+            ArrayBuilder<LocalSymbol> locals = ArrayBuilder<LocalSymbol>.GetInstance();
+            ArrayBuilder<BoundExpression> arguments = ArrayBuilder<BoundExpression>.GetInstance();
+
+            // left.HasValue && right.HasValue
+            BoundExpression condition = null;
+            foreach (var operand in operands)
+            {
+                if (operand.Type.IsNullableType())
+                {
+                    BoundExpression tempOperand = CaptureExpressionInTempIfNeeded(operand, sideeffects, locals);
+                    BoundExpression operandHasValue = MakeOptimizedHasValue(tempOperand.Syntax, tempOperand);
+
+                    if (condition is null)
+                    {
+                        condition = operandHasValue;
+                    }
+                    else
+                    {
+                        TypeSymbol boolType = _compilation.GetSpecialType(SpecialType.System_Boolean);
+                        condition = MakeBinaryOperator(syntax, BinaryOperatorKind.BoolAnd, condition, operandHasValue, boolType, method: null);
+                    }
+
+                    arguments.Add(MakeOptimizedGetValueOrDefault(tempOperand.Syntax, tempOperand));
+                }
+                else
+                {
+                    arguments.Add(operand);
```

Comment:
>arguments.Add(operand); [](start = 20, length = 23)

It looks like we might be changing order of evaluation here.

Reply:
Can you please clarify?


---
In reply to: [185367088](https://github.com/dotnet/roslyn/pull/26405#discussion_r185367088) [](ancestors = 185367088)

Reply:
> Can you please clarify?

Imagine the first operand is not nullable but the second is. It looks like the effect of this loop will be that the second operand is evaluated before the first one.

---
In reply to: [185370317](https://github.com/dotnet/roslyn/pull/26405#discussion_r185370317) [](ancestors = 185370317,185367088)

Reply:
so it is the case of:
index1..index2?
the order will be:
1) index1
2) index2.GetValueOrDefault()
for the arguments to `Range.Create()`
As for the `condition`, it will just be `index2.HasValue`


---
In reply to: [185382416](https://github.com/dotnet/roslyn/pull/26405#discussion_r185382416) [](ancestors = 185382416,185370317,185367088)

Reply:
Please check `LowerRange_Create_Index_NullableIndex()` in the emit tests.

---
In reply to: [185616205](https://github.com/dotnet/roslyn/pull/26405#discussion_r185616205) [](ancestors = 185616205,185382416,185370317,185367088)

Reply:
It is possible I am missing something here, but it also looks like This operand might not be evaluated at all when the other operand is null. I think both operands should be evaluated first and then lifting should happen

---
In reply to: [185616717](https://github.com/dotnet/roslyn/pull/26405#discussion_r185616717) [](ancestors = 185616717,185616205,185382416,185370317,185367088)